### PR TITLE
boilerplate login prompt

### PR DESCRIPTION
**Pull request by chenshay**

### CODE REVIEW

File: src/runtime/link-accounts-flow.js [hunk 1]
```diff
@@ -286,3 +286,52 @@ export class LinkSaveFlow {
...
+    /** @private @const {!Window} */
+    this.win_ = deps.win();
+
+    /** @private @const {!web-activities/activity-ports.ActivityPorts} */
+    this.activityPorts_ = deps.activities();
+
+    /** @private @const {!../components/dialog-manager.DialogManager} */
+    this.dialogManager_ = deps.dialogManager();
+
+    /** @private {!../api/subscriptions.AutoLoginRequestCallback} */
```
Comment (sohanirao) [inline src/runtime/link-accounts-flow.js]: indent

File: src/api/subscriptions.js [hunk 2]
```diff
@@ -219,6 +226,15 @@ export let OffersRequest;
 export let LoginRequest;
 
 
+/**
+ * @typedef {{
+ *   token: (string|undefined),
+ *   authCode: (string|undefined),
+ * }}
+ */
+export let AutoLoginRequestCallback;
```
Comment (sohanirao) [inline src/api/subscriptions.js]: Do you want to call this a callback if its just a value? Why not use AutoLoginRequest.

File: src/api/subscriptions.js [hunk 3]
```diff
@@ -219,6 +226,15 @@ export let OffersRequest;
 export let LoginRequest;
 
 
+/**
+ * @typedef {{
+ *   token: (string|undefined),
```
Comment (sohanirao) [inline src/api/subscriptions.js]: remove line 231-232 and replace with 'autoLoginHint: (string|undefined)' instead?

File: src/runtime/runtime-test.js [hunk 4]
```diff
@@ -1089,6 +1110,21 @@ describes.realWin('ConfiguredRuntime', {}, env => {
...
   });
 
+  it.only('should start autoLoginFlow', () => {
+    let autoLoginFlow;
+    const newPromise = new Promise(() => {});
+    sandbox.stub(AutoLoginFlow.prototype, 'start', function() {
+      autoLoginFlow = this;
+      return newPromise;
+    });
+    const resultPromise = runtime.autoLogin({token: 'test'});
```
Comment (sohanirao) [inline src/runtime/runtime-test.js]: loginHint instead

File: src/runtime/runtime-test.js [hunk 5]
```diff
@@ -645,6 +646,26 @@ describes.realWin('Runtime', {}, env => {
...
+      runtime.init('pub123');
+      const args = {
+        loginHint: 'hint123',
+        productId: 'pid123',
+      };
+
+      const newPromise = new Promise(() => {});
+      configuredRuntimeMock.expects('autoLogin').once();
+
+      const resultPromise = runtime.saveSubscription(args)
```
Comment (sohanirao) [inline src/runtime/runtime-test.js]: autoLogin API instead

**Comment by dvoytenko** [comment]
Not reviewing yet. But I think we need to quickly confirm the API shape and name via a minimal design doc.

File: src/api/subscriptions.js (hunk 3 above)
Comment (chenshay) [reply src/api/subscriptions.js]: done

File: src/api/subscriptions.js (hunk 2 above)
Comment (chenshay) [reply src/api/subscriptions.js]: done

File: src/runtime/link-accounts-flow.js (hunk 1 above)
Comment (chenshay) [reply src/runtime/link-accounts-flow.js]: done

File: src/runtime/runtime-test.js (hunk 5 above)
Comment (chenshay) [reply src/runtime/runtime-test.js]: done

File: src/runtime/runtime-test.js (hunk 4 above)
Comment (chenshay) [reply src/runtime/runtime-test.js]: not loginHint but isConsentRequired

File: docs/link-flow.md [hunk 6]
```diff
@@ -46,3 +46,10 @@ subscriptions.saveSubscription(() => requestPromise));
 
 The dialog will prompt the user to opt-in to save the subscription. If the user agrees, the provided callback will be called to resolve the token/authCode. The resulting promise will be resolved once the subscription has been saved. 
 
+# SwG Login Prompt flow
```
Comment (dvoytenko) [inline docs/link-flow.md]: We should either remove this section for now, or add "(not launched)" suffix.

File: docs/link-flow.md [hunk 7]
```diff
@@ -46,3 +46,10 @@ subscriptions.saveSubscription(() => requestPromise));
...
+
+This is the flow in which Google will ask the user for permission to log them in to the publisher's site. It is initiated by the publisher when the publisher doesn't find a user's subscription but Google does find the subscription.
+
+To enable this flow, 
+
+// TODO: chenshay finish this part.
```
Comment (dvoytenko) [inline docs/link-flow.md]: `TODO(username): ...`

File: src/api/subscriptions.js [hunk 8]
```diff
@@ -116,6 +116,12 @@ export class Subscriptions {
    */
   setOnLoginRequest(callback) {}
 
+  /**
+   * Starts the LoginPrompt flow.
+   * @return {!Promise}
+   */
+  loginPrompt() {}
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: We are currently using verbs for action methods. So, either `promptLogin` or `showLoginPrompt`?

File: src/runtime/link-accounts-flow-test.js [hunk 9]
```diff
@@ -24,11 +24,13 @@ import {
   LinkCompleteFlow,
   LinkbackFlow,
   LinkSaveFlow,
+  LoginPromptFlow,
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow-test.js]: It should be in its own module: this feature is pretty far away from linking.

File: src/runtime/link-accounts-flow.js [hunk 10]
```diff
@@ -286,3 +286,62 @@ export class LinkSaveFlow {
         });
   }
 }
+
+export class LoginPromptFlow {
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow.js]: Ditto: separate module needed.

File: src/runtime/link-accounts-flow.js [hunk 11]
```diff
@@ -286,3 +286,62 @@ export class LinkSaveFlow {
...
+    this.openViewPromise_ = null;
+
+    /** @private @const {!ActivityIframeView} */
+    this.activityIframeView_ = new ActivityIframeView(
+        this.win_,
+        this.activityPorts_,
+        feUrl('/loginpromptiframe'),
+        feArgs({
+          publicationId: deps.pageConfig().getPublicationId(),
+          productId: deps.pageConfig().getProductId(),
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow.js]: Possibly this is coming later if this is just a boilerplate. But heads up: we will need to pass `entitlements` value here. `deferred-account-flow.js` is probably a better model for this feature as a boilerplate. Feel free to keep and change it later, of course.

File: src/runtime/link-accounts-flow.js [hunk 12]
```diff
@@ -286,3 +286,62 @@ export class LinkSaveFlow {
...
+        }),
+        /* shouldFadeBody */ false
```
Comment (dvoytenko) [inline src/runtime/link-accounts-flow.js]: As a prompt, I'd think we need to fade body, so this should be `true`. Same as above: I think `deferred-account-flow.js` is a much closer model here.

File: src/api/subscriptions.js [hunk 13]
```diff
@@ -116,6 +116,12 @@ export class Subscriptions {
    */
   setOnLoginRequest(callback) {}
 
+  /**
+   * Starts the LoginPrompt flow.
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: Is the "waiting" API coming in a separate PR?

Comment (chenshay) [reply src/api/subscriptions.js]: Yes

Comment (chenshay) [reply src/api/subscriptions.js]: yes

File: docs/link-flow.md (hunk 6 above)
Comment (chenshay) [reply docs/link-flow.md]: done

File: docs/link-flow.md (hunk 7 above)
Comment (chenshay) [reply docs/link-flow.md]: done

File: src/api/subscriptions.js (hunk 8 above)
Comment (chenshay) [reply src/api/subscriptions.js]: done

File: src/runtime/link-accounts-flow-test.js (hunk 9 above)
Comment (chenshay) [reply src/runtime/link-accounts-flow-test.js]: done

File: src/runtime/link-accounts-flow.js (hunk 10 above)
Comment (chenshay) [reply src/runtime/link-accounts-flow.js]: done

File: src/runtime/link-accounts-flow.js (hunk 11 above)
Comment (chenshay) [reply src/runtime/link-accounts-flow.js]: done

File: src/runtime/link-accounts-flow.js (hunk 12 above)
Comment (chenshay) [reply src/runtime/link-accounts-flow.js]: done

File: src/runtime/runtime.js [hunk 14]
```diff
@@ -329,6 +332,14 @@ export class Runtime {
         });
   }
 
+  /** @override */
+  showLoginPrompt() {
+    return this.configured_(true)
+        .then(runtime => {
+          runtime.showLoginPrompt();
```
Comment (dvoytenko) [inline src/runtime/runtime.js]: You lost `return` here, looks like. And a unit test should have caught this.

File: src/api/subscriptions.js [hunk 15]
```diff
@@ -194,6 +200,12 @@ export const SubscriptionFlows = {
 };
 
 
+/** @enum {string} */
+export const LoginFlows = {
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: Let's move to the `SubscriptionFlows` enum - they idea is to keep all flow names in one enum.

File: src/runtime/runtime.js (hunk 14 above)
Comment (chenshay) [reply src/runtime/runtime.js]: done

File: src/api/subscriptions.js (hunk 15 above)
Comment (chenshay) [reply src/api/subscriptions.js]: done

**Pull request merged**